File: sol/solution/0001-0100/0034. Find First and Last Position of Element in Sorted Array/Solution.ex
defmodule Solution do
  @spec search_range(nums :: [integer], target :: integer) :: [integer]
  def search_range(nums, target) do
    [find_bound(nums, target, true), find_bound(nums, target, false)]
  end
  defp find_bound(nums, target, first) do
    helper(nums, target, 0, length(nums) - 1, -1, first)
  end
  defp helper(nums, target, l, r, ans, first) when l <= r do
    mid = div(l + r, 2)
    val = Enum.at(nums, mid)
    cond do
      val == target ->
        if first,
          do: helper(nums, target, l, mid - 1, mid, first),
          else: helper(nums, target, mid + 1, r, mid, first)
      val < target -> helper(nums, target, mid + 1, r, ans, first)
      true -> helper(nums, target, l, mid - 1, ans, first)
    end
  end
  defp helper(_, _, _, _, ans, _), do: ans
end

File: sol/solution/0001-0100/0034. Find First and Last Position of Element in Sorted Array/Solution.erl
-spec search_range([integer()], integer()) -> [integer()].
search_range(Nums, Target) ->
    {L, R} = {find_bound(Nums, Target, true), find_bound(Nums, Target, false)},
    [L, R].
find_bound(Nums, Target, First) ->
    loop(Nums, Target, 0, length(Nums)-1, -1, First).
loop(Nums, Target, L, R, Ans, First) when L =< R ->
    Mid = (L + R) div 2,
    Val = lists:nth(Mid+1, Nums),
    if Val =:= Target ->
            if First -> loop(Nums, Target, L, Mid-1, Mid, First);
               true -> loop(Nums, Target, Mid+1, R, Mid, First)
            end;
       Val < Target -> loop(Nums, Target, Mid+1, R, Ans, First);
       true -> loop(Nums, Target, L, Mid-1, Ans, First)
    end;
loop(_, _, _, _, Ans, _) -> Ans.